### PR TITLE
Show number of differences in the Compare editor #504

### PR DESCRIPTION
**Pull request by lathapatil**
Show number of differences (Ex: 6 Differences) in the toolbar of compare editor
which matches "Change markers" next to the scroll bar in compare editor

Fixes https://github.com/eclipse-platform/eclipse.platform/issues/504

Tested only in Linux and Windows OS

### CODE REVIEW

**Comment by iloveeclipse** [comment]
@lathapatil : could you please avoid *merge* commits? We don't allow them to be merged into master.

**Comment by iloveeclipse** [comment]
@lathapatil : you can see that you have now 3 commits in total, with two of them being merge commits.

Please locally (not in github, but in your IDE) create new branch based on latest master, cherry pick a **single** commit with your changes and force push to `Enhancements/504_Show_differences_in_compareEditor` branch.

**Comment by lathapatil** [comment]
> @lathapatil : could you please avoid _merge_ commits? We don't allow them to be merged into master.

I do realize now that "Update branch" button in this PR is actually doing _merge commits_ and instead I should have used _Sync fork_ and pull with rebase in my local then force pushed the branch.

**Comment by lathapatil** [comment]
> @lathapatil : you can see that you have now 3 commits in total, with two of them being merge commits.
> 
> Please locally (not in github, but in your IDE) create new branch based on latest master, cherry pick a **single** commit with your changes and force push to `Enhancements/504_Show_differences_in_compareEditor` branch.

Thanks for these steps , I was about to ask how for them as how to make it correct .

**Comment by lathapatil** [comment]
@iloveeclipse Could you please verify if commits are fine now and also review the PR ?

**Comment by iloveeclipse** [comment]
Commit looks good, but there is one related test fail, see https://ci.eclipse.org/platform/job/eclipse.platform/job/PR-731/4/testReport/junit/org.eclipse.compare.tests/TextMergeViewerTest/testToolbarLabelContribution/

```
java.lang.NullPointerException: Cannot invoke "org.eclipse.swt.widgets.Label.getText()" because the return value of "org.eclipse.compare.LabelContributionItem.getToolbarLabel()" is null
	at org.eclipse.compare.tests.TextMergeViewerTest.runInDialogWithToolbarDiffLabel(TextMergeViewerTest.java:530)
	at org.eclipse.compare.tests.TextMergeViewerTest.testToolbarLabelContribution(TextMergeViewerTest.java:501)
```

**Comment by lathapatil** [comment]
> continuous-integration/jenkins/pr-head



> one related test fail,

Actually this testcase is not failing in my local workspace . I am wondering how should I debug this and fix ? Can you suggest further on how do I get this setup where the testcase is failing..

**Comment by iloveeclipse** [comment]
> Actually this testcase is not failing in my local workspace . 

I can reproduce it. I'm on RHEL 7, Java 17, latest platform SDK build. What is your config?

> I am wondering how should I debug this and fix ? Can you suggest further on how do I get this setup where the testcase is failing..

Looking at the code, it calls some tasks asynchronously, so I assume you could get the test failing on a different system / with more/less cores enabled (have no idea about your setup).

Anyway, please apply the patch here and the test should pass:

```
diff --git a/team/tests/org.eclipse.compare.tests/src/org/eclipse/compare/tests/TextMergeViewerTest.java b/team/tests/org.eclipse.compare.tests/src/org/eclipse/compare/tests/TextMergeViewerTest.java
index a513593..9b3fd1a 100644
--- a/team/tests/org.eclipse.compare.tests/src/org/eclipse/compare/tests/TextMergeViewerTest.java
+++ b/team/tests/org.eclipse.compare.tests/src/org/eclipse/compare/tests/TextMergeViewerTest.java
@@ -32,3 +32,2 @@
 import org.eclipse.core.runtime.IPath;
-import java.net.URL;
 import org.eclipse.core.runtime.NullProgressMonitor;
@@ -526,11 +525,9 @@
 
-			IContributionItem contributionItem = toolbarManager.find("DiffCount");
-				assertNotNull(contributionItem);
-				LabelContributionItem labelContributionItem=(LabelContributionItem) contributionItem;
-				assertTrue(labelContributionItem.getToolbarLabel().getText().equals("7 Differences"));
-		try {
-			runnable.run();
-		} catch (WrappedException e) {
-			e.throwException();
-		}
+		processQueuedEvents();
+
+		IContributionItem contributionItem = toolbarManager.find("DiffCount");
+		assertNotNull(contributionItem);
+		LabelContributionItem labelContributionItem=(LabelContributionItem) contributionItem;
+		assertTrue(labelContributionItem.getToolbarLabel().getText().equals("7 Differences"));
+		runnable.run();
 		dialog.close();
@@ -539,2 +536,8 @@
 
+	private void processQueuedEvents() {
+		while (Display.getCurrent().readAndDispatch()) {
+			// Process all the events in the queue
+		}
+	}
+
 	private void runInDialogWithPartioner(Object input, Runnable runnable, final CompareConfiguration cc) throws Exception {
```

**Comment by lathapatil** [comment]
> I can reproduce it. I'm on RHEL 7, Java 17, latest platform SDK build. What is your config?

Thanks for identifying the issue further and the fix .

 I am working in Windows 10 , Java 17 and I do realize that testcase is passing in below platform build

 Eclipse Platform	4.30.0.v20230927-1800

When I updated yesterday to latest one , it is failing now with the same error

Eclipse IDE for Eclipse Committers (includes Incubating components)
Version: 2023-12 M1 (4.30.0 M1)
Build id: 20231005-1720

**Comment by iloveeclipse** [comment]
I've quickly tested the feature. In general, works a s expected, but it has a bug.
1) The label is not recomputed if the diff config changes.
2) The label is incorrect for zero changes - it says `0 Difference`

To reproduce: 
1) commit a change that only contains some spaces or tabs added.
2) enable "ignore whitespace" 
image
3) Diff - editor opens, shows  `0 Difference`
image
4) Toggle "Ignore whitespace" button on the main toolbar on/off - the label is unchanged but should reflect the diff state

**Comment by lathapatil** [comment]
> 3\. Diff - editor opens, shows  `0 Difference`

@iloveeclipse Thanks for the detailed test .
What is the expected display of the label here ? "No changes" or no Display is required when 0 difference is there ?

**Comment by iloveeclipse** [comment]
> What is the expected display of the label here ? "No changes" or no Display is required when 0 difference is there ?

"No changes" sounds good.

**Comment by lathapatil** [comment]
> 4\. Toggle "Ignore whitespace" button on the main toolbar on/off - the label is unchanged but should reflect the diff state

I will check and fix this bug !

**Comment by iloveeclipse** [comment]
> I will check and fix this bug !

Please also note that one can do edits in the compare editor and change number of diffy **without** triggering "whitespace" button.

**Comment by lathapatil** [comment]
> > I will check and fix this bug !
> 
> Please also note that one can do edits in the compare editor and change number of diffs **without** triggering "whitespace" button.

If I am not wrong this (Edit in compare editor and save which updates diff Label) is working fine. Only on trigger of "Ignore white space" Label is not getting updated

**Comment by iloveeclipse** [comment]
Test fail due https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1463 and https://github.com/eclipse-equinox/equinox/pull/382. 

Once that is sorted out, we can rebase & see if that helps.

**Comment by iloveeclipse** [comment]
The test fails as before, for same reason.
https://ci.eclipse.org/platform/job/eclipse.platform/job/PR-731/7/testReport/org.eclipse.compare.tests/TextMergeViewerTest/testToolbarLabelContribution/

Please try to adopt test as proposed above.

**Comment by lathapatil** [comment]
> > I will check and fix this bug !
> 
> Please also note that one can do edits in the compare editor and change number of diffs **without** triggering "whitespace" button.

@iloveeclipse I have fixed the points received in the review . Let me know if anything is still not working out in this feature.

**Comment by iloveeclipse** [comment]
The errors are coming from tycho eclipse-tycho/tycho#3019

**Comment by lathapatil** [comment]
> The errors are coming from tycho [eclipse-tycho/tycho#3019](https://github.com/eclipse-tycho/tycho/issues/3019)

Is this issue resolved ? If yes, how can I retrigger the build to see latest status?

**Comment by HeikoKlare** [comment]
@lathapatil The build can now be retriggered. You only have to manually rebase your changes on current master branch, as the version bump for `org.eclipse.compare` in this PR became conflicting with the change in the master branch.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java
Comment (iloveeclipse) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:5433]: Some issues: 
1) display should be better retrieved from the widget on which we are operating
2) in all `asyncExec()` cals first thing to do is to check if the UI we are referring to is still there.
3) `setRedraw(true)` is wrong. I guess you wanted to use `control.getParent().redraw();` but also this is questionable & I don't see a need for this.

So please replace this with
```
			fComposite.getDisplay().asyncExec(() -> {
				// relayout in next tick
				ToolBar control = tbm.getControl();
				if (control != null && !control.isDisposed()) {
					tbm.update(true);
				}
			});
```

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java
Comment (iloveeclipse) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: Few issues here:
1) `SWT.LEFT` used in constructors of COmposite & Label  don't make any sense as the argument is not alignment => `SWT.NONE`
2) hard coded margins looks strange => remove
3) I would not align label on bottom, it looks odd as it is not aligned with buttons anymore => `SWT.CENTER`

With that, please try
```
	@Override
	protected Control createControl(Composite parent) {
		Composite composite = new Composite(parent, SWT.NONE);
		this.toolbarLabel = new Label(composite, SWT.NONE);

		GridLayout compositeLayout = new GridLayout(1, false);
		composite.setLayout(compositeLayout);
		GridData labelLayout = new GridData(SWT.LEFT, SWT.CENTER, true, true);

		this.toolbarLabel.setLayoutData(labelLayout);
		this.toolbarLabel.setText(this.labelName);
		return composite;
	}
```

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
Thanks, I've tested locally, all the changes are now properly reflected in the diff label.

I've reviewed the rest, please fix the two issues I've commented on and we are ready to merge after that.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java
Comment (lathapatil) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:5433]: To be in sync the below code was taken from InternalRefresh Method of ContentMergeViewer.java 

`Display.getDefault().asyncExec(() -> {`
						`// relayout in next tick`
						`tbm.update(true);`
						`tbm.getControl().getParent().setRedraw(true);`
					`});`
which was fixed for "bad UI on MacOS merge viewer" . Should it be replaced with below code here now as it is changed there ?
`tbm.update(true);`
`tbm.getControl().getParent().layout(true);`

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java
Comment (lathapatil) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: > 2\. hard coded margins looks strange => remove

If we remove these the label looks like below (Not aligned properly)

diffLabel

Comment (iloveeclipse) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: > > 2. hard coded margins looks strange => remove
> 
> If we remove these the label looks like below (Not aligned properly)
> 
> diffLabel

Which OS is it?

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java
Comment (iloveeclipse) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java:5433]: Please check if proposed code works on Windows. If yes, no need to copy/paste code you mention, especially we know it doesn't work properly on Mac (this is the reason it is changed back and forth).

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java
Comment (iloveeclipse) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: In that case I would only enable Windows specific workaroumds in "if(isWindows)" blocks.

Comment (lathapatil) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: > In that case I would only enable Windows specific workaroumds in "if(isWindows)" blocks.

Have you verified on Linux and Mac already if this is working without margin values ?

Comment (iloveeclipse) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: I've verified on Linux, I have no Mac access.

Comment (lathapatil) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java:48]: > 3\. I would not align label on bottom, it looks odd as it is not aligned with buttons anymore => `SWT.CENTER`

Here , I observed that label appearance does not change whatever SWT value I use (tested in Ubuntu and Windows). Where can we see the impact ?

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java
Comment (lathapatil) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/contentmergeviewer/TextMergeViewer.java]: The changes are adopted . Please have a look

**Review by iloveeclipse — APPROVED** [review]
Thanks for your patience & contribution. Looks good now.

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java
Comment (pbodnar) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/LabelContributionItem.java]: > In that case I would only enable Windows specific workaroumds in "if(isWindows)" blocks.

I believe this "special handling of margins on Windows" should be better also commented in-line in the code, as it will most probably confuse someone sooner or later. Anyway, thank you both for a good job on this PR! 👍

**Pull request merged**
